app/src/main/java/com/infomaniak/swisstransfer/upload/TransferUploader.kt

```diff
@@ -199,101 +199,87 @@ class TransferUploader(
+                requestSemaphore.withPermit {
+                    uploadChunkIfNeeded(metadata = metadata, chunkIndex = chunkIndex, isLastChunk = false)
                 }
             }
         }
-
-        byteArrayPool.clear()
+        uploadChunkIfNeeded(metadata = metadata, chunkIndex = lastChunkIndex, isLastChunk = true)
     }
 
-    private fun getReusableByteArray(
-        byteArrayPool: ArrayBlockingQueue<ByteArray>,
-        inputStream: BufferedInputStream,
-        chunkSize: Long,
-        isLastChunk: Boolean,
-    ): ByteArray {
-        val currentChunkSize = if (isLastChunk) inputStream.available() else chunkSize.toInt()
-        val reusableByteArray = if (isLastChunk) null else byteArrayPool.poll()
+    private suspend fun uploadChunkIfNeeded(metadata: FileToUploadMetaData, chunkIndex: Int, isLastChunk: Boolean) {
+        val uploadStatus = metadata.chunksUploadStatus[chunkIndex]
+        val fileUuid = metadata.uuid
+        if (uploadStatus == DefinitelyComplete) {
+            SentryLog.d(TAG, "skipping chunk #$chunkIndex since it's already been uploaded (file $fileUuid)")
+            return
+        }
+        val targetFileUri = metadata.pickedFile.uri
+
+        val chunkSize = metadata.chunkConfig.fileChunkSize
+        val totalFileSize = metadata.pickedFile.size
+        val content = targetFileUri.toContentChunk(
+            offset = chunkSize * chunkIndex,
+            length = if (isLastChunk) {
+                (totalFileSize % chunkSize).let { if (it == 0L) chunkSize.toLong() else it }
+            } else chunkSize
+        )
+        metadata.chunksUploadStatus[chunkIndex] = StartedOrComplete
+        withRetries {
+            SentryLog.i(TAG, "Uploading chunk #$chunkIndex of $fileUuid") {
+                uploadChunk(
+                    fileUUID = fileUuid,
+                    chunkIndex = chunkIndex,
+                    isLastChunk = isLastChunk,
+                    data = content
+                )
+            }
+        }
+        metadata.chunksUploadStatus[chunkIndex] = DefinitelyComplete
+    }
 
-        return reusableByteArray ?: ByteArray(currentChunkSize)
+    private suspend inline fun <R> withRetries(block: () -> R): R {
+        val giveUpDelay = 40.seconds
+        val maxDelayBetweenRetries = 10.seconds
+        val minDelayBetweenRetries = 500.milliseconds
+        var attemptNumber = 0
+        val attemptTimeMark = TimeSource.Monotonic.markNow()
+        while (true) {
+            try {
+                attemptNumber++
+                return block()
+            } catch (e: NetworkException) {
+                val timeUntilGiveUp = giveUpDelay - attemptTimeMark.elapsedNow()
+                if (timeUntilGiveUp < Duration.ZERO) throw e
+                val retryIn = (minDelayBetweenRetries * attemptNumber)
+                    .coerceAtMost(maxDelayBetweenRetries)
+                    .coerceAtMost(timeUntilGiveUp)
```

Comment:
I'm not sure this second coerceAtMost makes a lot of sense. 
We could juste let the maxDelayBetween retry even if it has for example only 4s remaining out of the 40. 
It seems a bit weird to me to increase a delay, then the last time retry faster

Reply:
The idea is that the last retry can start/occur up to the "give up delay" (40s at the moment), so that the max waiting time until give us is consistent.

I find tracking what the algorithm can do to be easier with that delay "clamping".

Reply:
This add more reflexion to know what this does and be sure to understand why it was placed here.
We should just remove it to simplify

Reply:
This add more reflexion to know what this does and be sure to understand why it was placed here.
We should just remove it to simplify

Reply:
The intent is exactly the opposite though. I added a clarifying comment.

Reply:
Changed the algorithm to have a max delay instead, as seen offline.